docs/book/src/tasks/experimental-features/runtime-sdk/implement-topology-mutation-hook.md

```diff
@@ -0,0 +1,177 @@
+  be retrieved, but be aware of performance impact.
+* **Deterministic results**: For a given request (a set of templates and variables) an External Patch Extension must
+  always return the same response (a set of patches). Otherwise the Cluster topology will never reach a stable state.
+* **Idempotence**: An External Patch Extension must only return patches if changes to the templates are required,
+  i.e. unnecessary patches when the template is already in the desired state must be avoided.
+* **Avoid Dependencies**: An External Patch Extension must be independent of other External Patch Extensions. However
+  if dependencies cannot be avoided, it is possible to control the order in which patches are executed via the ClusterClass.
+
+## Definitions
+
+### GeneratePatches
+
+A GeneratePatches call generates patches for the entire Cluster topology. Accordingly the request contains all
+templates, the global variables and the template-specific variables. The response contains generated patches.
+
+#### Example request:
+
+* Generating patches for a Cluster topology is done via a single call to allow External Patch Extensions a
+  holistic view of the entire Cluster topology. Additionally this allows us to reduce the number of round-trips.
+* Each item in the request will contain the template as a raw object. Additionally information about where
+  the template is used is provided via `holderReference`.
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: GeneratePatchesRequest
+variables:
+- name: <variable-name>
+  value: <variable-value>
+  ...
+items:
+- uid: 7091de79-e26c-4af5-8be3-071bc4b102c9
+  holderReference:
+    apiVersion: cluster.x-k8s.io/v1beta1
+    kind: MachineDeployment
+    namespace: default
+    name: cluster-md1-xyz
+    fieldPath: spec.template.spec.infrastructureRef
+  object:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+    kind: AWSMachineTemplate
+    spec:
+    ...
+  variables:
+  - name: <variable-name>
+    value: <variable-value>
+    ...
+```
+
+#### Example Response:
+
+* The response contains patches instead of full objects to reduce the payload.
+* Templates in the request and patches in the response will be correlated via UIDs.
+* Like inline patches, external patches are only allowed to change fields in `spec.template.spec`.
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: GeneratePatchesResponse
+status: Success # or Failure
+message: "error message if status == Failure"
+items:
+- uid: 7091de79-e26c-4af5-8be3-071bc4b102c9
+  patchType: JSONPatch
+  patch: <JSON-patch>
+```
+
+For additional details, refer to the [Draft OpenAPI spec](https://editor.swagger.io/?url=https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/docs/proposals/images/topology-mutation-hook/runtime-sdk-openapi.yaml).
+
+We are considering introducing a library to facilitate development of External Patch Extensions. It would provide capabilities like:
```

Comment:
```suggestion
We are considering to introduce a library to facilitate development of External Patch Extensions. It would provide capabilities like:
```